connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -202,6 +204,225 @@ public Object valueFrom(Map<String, Object> map) {
+                            Struct fieldValue = updateValue(
+                                field.schema(),
+                                originalValue.getStruct(field.name()),
+                                updateSchema.field(field.name()).schema(),
+                                step + 1,
+                                whenFound,
+                                whenNotFound,
+                                whenOther
+                            );
+                            updated.put(field.name(), fieldValue);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            found = false;
+                            whenOther.apply(originalValue, field, updated, null, this);
+                        }
+                    }
+                } else {
+                    whenOther.apply(originalValue, field, updated, null, this);
+                }
+            }
+        }
+        if (!found) {
+            whenNotFound.apply(
+                originalValue,
+                null,
+                updated,
+                updateSchema.field(steps.get(step)),
+                this);
+        }
+        return updated;
+    }
+
+    /**
+     * Prepares a new schema based on an original one, and applies an update function
+     * when the current path(s) is found.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields are copied from original schema.
+     *
+     * @param originalSchema        baseline schema
+     * @param baselineSchemaBuilder baseline schema build, if changes to the baseline
+     *                              are required before copying original
+     * @param whenFound             function to apply when current path(s) is/are found.
+     * @return an updated schema. Resulting schemas are usually cached for further access.
+     */
+    public Schema updateSchemaFrom(
+        Schema originalSchema,
+        SchemaBuilder baselineSchemaBuilder,
+        StructSchemaUpdater whenFound,
+        StructSchemaUpdater whenNotFound,
+        StructSchemaUpdater whenOther
+    ) {
+        return updateSchema(originalSchema, baselineSchemaBuilder, 0, whenFound, whenNotFound, whenOther);
+    }
+
+    // Recursive implementation to update schema at different steps.
+    // Consider that resulting schemas are usually cached.
+    private Schema updateSchema(
+        Schema operatingSchema,
+        SchemaBuilder builder,
+        int step,
+        StructSchemaUpdater matching,
+        StructSchemaUpdater notFound,
+        StructSchemaUpdater others
+    ) {
+        if (operatingSchema.isOptional()) {
+            builder.optional();
+        }
+        if (operatingSchema.defaultValue() != null) {
+            builder.defaultValue(operatingSchema.defaultValue());
+        }
```

Comment:
Is this correct? What if the schema we're modifying is for a `Struct`? Wouldn't this cause the new schema to have a default `Struct` value that uses the old schema?

Reply:
Good catch. I'm removing this for the current PR -- we can evaluate if this is needed in following PRs.

Reply:
Hmmm... I think this is definitely required, otherwise we drop default values for nested structs whose children we modify.